### ISSUE TITLE
FR: JSON schema for Tailscale ACL

### ISSUE THREAD

**Issue by joneskoo** [tailscale/tailscale#10794]
### What are you trying to do?

I want my editor (VS Code) to validate syntax and provide completion for Tailscale ACL. I would like it to complete e.g. a new entry in grants, and provide description of expected values.

### How should we solve this?

I would like for Tailscale to provide an official JSON Schema document that can be used to allow VS Code and other editors that support JSON Schema to validate and provide structural hints.

### What is the impact of not solving this?

Iterating on ACL currently requires making a pull request and waiting for CI to run validation and provide feedback. By validating syntax using JSON schema locally in editor, this is simplified.

### Anything else?

While VS Code extension is not required to make use of this, and I believe the JSON schema would be useful for other editors like JetBrains IDEs, it would be useful that Tailscale makes it easy to use this with VS Code. Maybe the Tailscale VS Code extension should recognize Tailscale ACL file and define schema, or maybe it's possible to provide simple usage instructions that don't require any extension.

**Comment by joneskoo** [comment]
Calling a remote API for local editor validation sounds unusual to me, unless you're saying that code could be embedded in an extension?

If schema is hosted in a way where Tailscale can maintain it (remote https url), that should resolve your concern?

https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings

**Comment by joneskoo** [comment]
If you want to give it a try to see why this is useful, you can quickly test it out by adding property

```json
{
	"$schema": "./tailscale-acl.json-schema.json"
}
```

And adding to your workspace this incomplete draft of JSON schema:
[tailscale-acl.json-schema.json](https://github.com/tailscale/tailscale/files/13884127/tailscale-acl.json-schema.json)


<img width="929" alt="image" src="https://github.com/tailscale/tailscale/assets/305104/f61a9e20-507b-40dc-8a18-d6d1202b931d">

<img width="340" alt="image" src="https://github.com/tailscale/tailscale/assets/305104/d7d5f226-1894-4480-8b2f-7e9288002094">

Assuming you can host the JSON schema in a remote registry such that it doesn't need to be kept in sync by the user, wouldn't this be pretty big value? And even then, I wouldn't see having to update schema to a new version occasionally as a blocker for Tailscale providing the official schema.

**Comment by Preisschild** [comment]
Also the gitops-pusher complains `json: unknown field "$schema"`  when adding a schema reference

**Comment by Preisschild** [comment]
> There is an API endpoint which might be useful: [`main`/api.md#validate-and-test-policy-file](https://github.com/tailscale/tailscale/blob/main/api.md#validate-and-test-policy-file) The JSON format evolves frequently enough that we are concerned about the support load of people using a snapshot of a schema saved at a point in time.

The schema could be just be available via an unauthenticated `GET` request over the tailscale API.


```json
	"$schema": "https://api.tailscale.com/api/v2/acl.schema.json"
	"hosts": {
	}
```

would work fine

**Comment by Radiergummi** [comment]
Yeah, seconded. This would be an amazing, low-maintenance but high-value feature. Allow (and encourage) a `$schema` property on the policy, serve that from your API, and every GitOps user gets automatic and exhaustive autocompletion for the Tailscale policy in almost any modern code editor or IDE out of the box.

The only thing you'd need to do is define the schema and update it when you make changes, but AI agents are *really* good at this by now.

**Comment by joneskoo** [comment]
I prototyped an implementation of this using the existing `<tailscale.com/client/tailscale/v2`> types — no new structs to maintain. The schema updates automatically when the dependency is bumped and `go generate` is re-run.

**Code:** [`cmd/gen-acl-schema/main.go`](https://github.com/joneskoo/tailscale/blob/claude/tailscale-acl-json-schema-v2/cmd/gen-acl-schema/main.go) + `main_test.go`
**Generated schema:** [`acl-schema.json`](https://raw.githubusercontent.com/joneskoo/tailscale/refs/heads/claude/tailscale-acl-json-schema-v2/acl-schema.json)

PoC branch [joneskoo/tailscale branch claude/tailscale-acl-json-schema-v2](https://github.com/joneskoo/tailscale/tree/claude/tailscale-acl-json-schema-v2)

The generator was produced with Claude Code. I asked for a prompt to specify how it was produced, shared so maintainers can reproduce or adapt it with their own LLM.

Let me known if Claude created PR like this is accepted or would you rather adjust according to Tailscale conventions - probably better without PR. Shared as reference and because this schema is already usable and passes both test data and my policies.

---

> Create a `cmd/gen-acl-schema` tool in the tailscale/tailscale repo that generates `acl-schema.json` (JSON Schema 2020-12) for the Tailscale ACL policy file format.
>
> Requirements:
> - Use `<tailscale.com/client/tailscale/v2.ACL`> as the source struct — it's already a module dependency, so the schema stays in sync when the dep is bumped. No new typed packages.
> - Add a `//go:generate go run <tailscale.com/cmd/gen-acl-schema> -o ../../acl-schema.json` directive so the schema is regenerated with one command.
> - Use `<github.com/invopop/jsonschema`> for reflection. Set `RequiredFromJSONSchemaTags: true`.
> - Extract doc comments from the upstream source using `r.AddGoComments("<tailscale.com/client/tailscale/v2>", dir)` where `dir` comes from `go list -m -json`. Then overlay a hand-written `commentMap` (keyed `"pkg.Type.GoField"`) so hand-written descriptions take precedence over extracted ones. Source descriptions from <https://tailscale.com/kb/1337/policy-syntax>.
> - `SSHCheckPeriod` is a `time.Duration` alias that marshals as a string (`"20h"`, `"always"`). Add a `Reflector.Mapper` override to emit `{type: string}` for it instead of integer.
> - Write output with `<tailscale.com/atomicfile.WriteFile`>.
> - Add a test (`main_test.go`) that generates the schema to a temp file, compiles it with `<github.com/santhosh-tekuri/jsonschema/v6`>, then validates both `testdata/acl.json` and `testdata/acl.hujson` from the module cache. Standardize HuJSON with `<github.com/tailscale/hujson`> before validation.
> - Commit the generated `acl-schema.json` at the repo root.
> - Set `schema.ID = "<https://raw.githubusercontent.com/tailscale/tailscale/main/acl-schema.json>"`.
